reassembly/tcpassembly.go

```diff
@@ -0,0 +1,1319 @@
+// for handling TCP wrap-around.  The sequence should always be in the range
+// [0,0xFFFFFFFF]... its other bits are simply used in wrap-around calculations
+// and should never be set.
+type Sequence int64
+
+// Difference defines an ordering for comparing TCP sequences that's safe for
+// roll-overs.  It returns:
+//    > 0 : if t comes after s
+//    < 0 : if t comes before s
+//      0 : if t == s
+// The number returned is the sequence difference, so 4.Difference(8) will
+// return 4.
+//
+// It handles rollovers by considering any sequence in the first quarter of the
+// uint32 space to be after any sequence in the last quarter of that space, thus
+// wrapping the uint32 space.
+func (s Sequence) Difference(t Sequence) int {
+	if s > uint32Max-uint32Max/4 && t < uint32Max/4 {
+		t += uint32Max
+	} else if t > uint32Max-uint32Max/4 && s < uint32Max/4 {
+		s += uint32Max
+	}
+	return int(t - s)
+}
+
+// Add adds an integer to a sequence and returns the resulting sequence.
+func (s Sequence) Add(t int) Sequence {
+	return (s + Sequence(t)) & uint32Max
+}
+
+// TCPAssemblyStats provides some figures for a ScatterGather
+type TCPAssemblyStats struct {
+	// For this ScatterGather
+	Chunks  int
+	Packets int
+	// For the half connection, since last call to ReassembledSG()
+	QueuedBytes    int
+	QueuedPackets  int
+	OverlapBytes   int
+	OverlapPackets int
+}
+
+// ScatterGather is used to pass reassembled data and metadata of reassembled
+// packets to a Stream via ReassembledSG
+type ScatterGather interface {
+	// Returns the length of available bytes and saved bytes
+	Lengths() (int, int)
+	// Returns the bytes up to length
+	Fetch(length int) []byte
+	// Tell to keep from offset
+	KeepFrom(offset int)
+	// Return CaptureInfo of packet corresponding to given offset
+	CaptureInfo(offset int) gopacket.CaptureInfo
+	// Return some info about the reassembled chunks
+	Info() (direction TCPFlowDirection, start bool, end bool, skip int)
+	// Return some stats regarding the state of the stream
+	Stats() TCPAssemblyStats
+}
+
+// byteContainer is either a page or a livePacket
+type byteContainer interface {
+	Bytes() []byte
+	Length() int
+	ConvertToPages(*pageCache, int, AssemblerContext) (*page, *page, int)
+	CaptureInfo() gopacket.CaptureInfo
+	AssemblerContext() AssemblerContext
+	Release(*pageCache) int
+	Start() bool
+	End() bool
+	Seq() Sequence
+	IsPacket() bool
+}
+
+// Implements a ScatterGather
+type reassemblyObject struct {
+	all       []byteContainer
+	Skip      int
+	Direction TCPFlowDirection
+	saved     int
+	toKeep    int
+	// stats
+	queuedBytes    int
+	queuedPackets  int
+	overlapBytes   int
+	overlapPackets int
+}
+
+func (rl *reassemblyObject) Lengths() (int, int) {
+	l := 0
+	for _, r := range rl.all {
+		l += r.Length()
+	}
+	return l, rl.saved
+}
+
+func (rl *reassemblyObject) Fetch(l int) []byte {
+	if len(rl.all) == 1 && l == rl.all[0].Length() {
+		return rl.all[0].Bytes()
+	}
+	bytes := []byte{}
```

Comment:
It was hidden in the small TODO on the top:"- implement chunked (cheap) reads and Reader() interface".